model.py

```diff
@@ -374,38 +359,36 @@ def _make_program(self, inputs):
         self._endpoints[self.mode] = {
             "output": outputs,
             "loss": losses
+        } if self.model._loss_function else {
+            'output': outputs
         }
 
-    def _infer_input_vars(self, inputs):
-        input_vars = []
-        for idx, i in enumerate(inputs):
-            if i is None:  # train and test may take different arguments
-                input_vars.append(None)
-                continue
-            ndarray = np.array(i)
-            name = list(self._input_desc.keys())[idx]
-            shape = list(self._input_desc.values())[idx]
-            if shape is None:
-                shape = (None, ) + ndarray.shape[1:]
-            input_vars.append(fluid.data(name, shape, ndarray.dtype))
-        return input_vars
-
-    def _infer_label_vars(self, outputs):
-        shapes = self.model._loss_function.infer_shape(outputs)
-        dtypes = self.model._loss_function.infer_dtype(outputs)
-        label_vars = []
-        for idx, (shape, dtype) in enumerate(zip(shapes, dtypes)):
-            name = '__label{}'.format(idx)
-            label_vars.append(fluid.data(name, shape, dtype))
-        return label_vars
-
-    def _compile_and_initialize(self, prog, device='CPU', device_ids=None):
+    def prepare(self):
         compiled_prog = self._compiled_progs.get(self.mode, None)
         if compiled_prog is not None:
             return compiled_prog
 
-        places = [device.lower() == 'gpu' and fluid.CUDAPlace(i)
-                  or fluid.CPUPlace() for i in device_ids]
+        if isinstance(self.model._inputs, dict):
+            ins = [self.model._inputs[n] \
+                for n in extract_args(self.model.forward) if n != 'self']
+        else:
+            ins = self.model._inputs
+        self._input_vars[self.mode] = [k.forward() for k in to_list(ins)]
```

Comment:
It seems that `_input_vars` for different modes all has same contents since `model._inputs` can only  be set once in `model.prepare`.  However, it doesn't matter.

Reply:
Yeah.  see comments for `train和eval复用Model导致的缺点` in beginning of PR.